lib/DXIL/DxilPDB.cpp

```diff
@@ -0,0 +1,489 @@
+    return E_FAIL;
+
+  hlsl::DxilContainerHeader *DxilHeader = (hlsl::DxilContainerHeader *)pContainer->GetBufferPointer();
+  for (unsigned i = 0; i < DxilHeader->PartCount; i++) {
+    hlsl::DxilPartHeader *PartHeader = GetDxilContainerPart(DxilHeader, i);
+    if (PartHeader->PartFourCC == hlsl::DFCC_ShaderHash) {
+      hlsl::DxilShaderHash *HashHeader = (hlsl::DxilShaderHash *)(PartHeader+1);
+      if (sizeof(HashHeader->Digest) > uCapacity)
+        return E_FAIL;
+      memcpy(pDigest, HashHeader->Digest, sizeof(HashHeader->Digest));
+      *pBytesWritten = sizeof(HashHeader->Digest);
+      return S_OK;
+    }
+  }
+  return E_FAIL;
+}
+
+HRESULT hlsl::pdb::WriteDxilPDB(IMalloc *pMalloc, IDxcBlob *pContainer, IDxcBlob **ppOutBlob) {
+  if (!hlsl::IsValidDxilContainer((hlsl::DxilContainerHeader *)pContainer->GetBufferPointer(), pContainer->GetBufferSize()))
+    return E_FAIL;
+
+  BYTE HashData[16];
+  UINT32 uHashSize = 0;
+  IFR(FindShaderHash(pContainer, HashData, sizeof(HashData), &uHashSize));
+  SmallVector<char, 0> PdbStream = WritePdbStream({ HashData, HashData+uHashSize });
+
+  MSFWriter Writer;
+  Writer.AddEmptyStream();     // Old Directory
+  Writer.AddStream(PdbStream); // PDB Header
+
+  // Fixed streams
+  Writer.AddEmptyStream(); // TPI
+  Writer.AddEmptyStream(); // DBI
+  Writer.AddEmptyStream(); // IPI
+  
+  Writer.AddStream({ (char *)pContainer->GetBufferPointer(), pContainer->GetBufferSize() }); // Actual data block
+  
+  CComPtr<hlsl::AbstractMemoryStream> pStream;
+  IFR(hlsl::CreateMemoryStream(pMalloc, &pStream));
+
+  raw_stream_ostream OS(pStream);
+  Writer.WriteToStream(OS);
+  OS.flush();
+
+  IFR(pStream.QueryInterface(ppOutBlob));
+
+  return S_OK;
+}
+
+
+struct PDBReader {
+  IStream *m_pStream = nullptr;
+  IMalloc *m_pMalloc = nullptr;
+  UINT32 m_uOriginalOffset = 0;
+  MSF_SuperBlock m_SB = {};
+  HRESULT m_Status = S_OK;
+
+  HRESULT SetPostion(INT32 sOffset) {
+    LARGE_INTEGER Distance = {};
+    Distance.QuadPart = m_uOriginalOffset + sOffset;
+    ULARGE_INTEGER NewLocation = {};
+    return m_pStream->Seek(Distance, STREAM_SEEK_SET, &NewLocation);
+  }
+
+  PDBReader(IMalloc *pMalloc, IStream *pStream) : m_pStream(pStream), m_pMalloc(pMalloc) {
+    m_Status = ReadSuperblock(&m_SB);
+  }
+
+  // Reset the stream back to its original position, regardless of
+  // we succeeded or failed.
+  ~PDBReader() {
+    SetPostion(0);
+  }
+
+  HRESULT GetStatus() { return m_Status; }
+
+  HRESULT ReadSuperblock(MSF_SuperBlock *pSB) {
+    ULONG SizeRead = 0;
+    IFR(m_pStream->Read(pSB, sizeof(*pSB), &SizeRead));
+    if (memcmp(pSB->MagicBytes, kMsfMagic, sizeof(kMsfMagic)) != 0)
+      return E_FAIL;
+
+    return S_OK;
+  }
+
+  HRESULT ReadU32(UINT32 *pValue) {
+    ULONG NumBytesRead = 0;
+    support::ulittle32_t ValueLE;
+    IFR(m_pStream->Read(&ValueLE, sizeof(ValueLE), &NumBytesRead));
```

Comment:
Is it a concern that we always ignore the `numbytesread`? We could have a ReadAllBytes(IStream, void*, SIZE_T) that fails if the stream reaches its end and does not read everything.

Reply:
Good idea.